src/echo/echo.rs

```diff
@@ -36,119 +33,77 @@ static HELP: &str = r#"
  \\xHH    byte with hexadecimal value HH (1 to 2 digits)
 "#;
 
-enum Base {
-    B8 = 8,
-    B16 = 16,
-}
-
-struct Opts {
-    newline: bool,
-    escape: bool,
-}
-
-fn convert_str(string: &[u8], index: usize, base: Base) -> (char, usize) {
-    let (max_digits, is_legal_digit): (usize, fn(u8) -> bool) = match base {
-        Base::B8 => (3, |c| (c as char).is_digit(8)),
-        Base::B16 => (2, |c| (c as char).is_digit(16)),
-    };
-
-    let mut bytes = vec![];
-    for offset in 0..max_digits {
-        if string.len() <= index + offset as usize {
-            break;
-        }
-        let c = string[index + offset as usize];
-        if is_legal_digit(c) {
-            bytes.push(c as u8);
-        } else {
-            break;
-        }
+pub fn unescape(input: &str) -> String {
+    let mut output = String::new();
+    let mut iter = input.chars();
+    while let Some(c) = iter.next() {
+        output.push(match c {
+            '\\' => match iter.next() {
+                Some(c) => match c {
+                    '\\' => '\\',
+                    'a' => '\x07',
+                    'b' => '\x08',
+                    'c' => break,
+                    'e' => '\x1b',
+                    'f' => '\x0c',
+                    'n' => '\n',
+                    'r' => '\r',
+                    't' => '\t',
+                    'v' => '\x0b',
+                    'x' => match (iter.next(), iter.next()) {
+                        (Some(hi), Some(lo)) => match (hi.to_digit(16), lo.to_digit(16)) {
+                            (Some(hi), Some(lo)) => std::char::from_u32((hi << 4) | lo).unwrap(),
+                            _ => continue,
```

Comment:
This behavior is incorrect.  `\x` takes *up to* 2 digits, not 2 digits or nothing.  Same for `\0` (*up to* 3 digits, not 3 digits or nothing).